### PR TITLE
8339733: C2: some nodes can have incorrect control after do_range_check()

### PR DESCRIPTION
**Pull request by rwestrel**
PhaseIdealLoop::do_range_check() sets the control of the new pre and
main limits to be the entry control of the pre loop but it eliminates
all conditions whose parameters are invariant in the main loop. Most
of the time they are also invariant in the pre loop but that's not
guaranteed. It does happen sometimes that those parameters are pinned
in the pre loop. In that case, PhaseIdealLoop::do_range_check() sets
wrong controls.

This doesn't cause any issue today AFAICT.

Also, this seems to be a typo in PhaseIdealLoop::insert_pre_post_loops():

```
pre_head->in(0)
```

is `pre_head`. I fixed that one too.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8339733](https://bugs.openjdk.org/browse/JDK-8339733): C2: some nodes can have incorrect control after do_range_check() (**Bug** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20908/head:pull/20908` \
`$ git checkout pull/20908`

Update a local copy of the PR: \
`$ git checkout pull/20908` \
`$ git pull https://git.openjdk.org/jdk.git pull/20908/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20908`

View PR using the GUI difftool: \
`$ git pr show -t 20908`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20908.diff">https://git.openjdk.org/jdk/pull/20908.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20908#issuecomment-2337470440)

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -3112,6 +3132,35 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
...
   _igvn.replace_input_of(opqzm, 1, main_limit);
+  // new main_limit can push Bool/Cmp nodes down (when one of the eliminated condition has parameters that are not loop
+  // invariant in the pre loop.
+  set_ctrl(opqzm, new_limit_ctrl);
+  set_ctrl(iffm->in(1), new_limit_ctrl);
+  set_ctrl(iffm->in(1)->in(1), new_limit_ctrl);
+}
+
+// Adjust control for node and its inputs (and inputs of its inputs) to be above the pre end
+void PhaseIdealLoop::ensure_node_and_inputs_are_above_pre_end(CountedLoopEndNode* pre_end, Node* node, Node*& control) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:3143]: `control` should be the `ctrl` of `node` here, right (i.e. `control == get_ctrl(node)`)? Maybe we can assert that. But thinking about it, would it hurt to just call `get_ctrl(node)` again here and remove `control` as parameter? Then we can just return the new control and do 
```
offset_c = ensure_node_and_inputs_are_above_pre_end(pre_end, offset);
```
at the call-site.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 2]
```diff
@@ -3072,8 +3086,14 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
...
     pre_limit = (stride_con > 0) ? (Node*)new MinINode(pre_limit, orig_limit)
                                  : (Node*)new MaxINode(pre_limit, orig_limit);
-    register_new_node(pre_limit, pre_ctrl);
+    register_new_node(pre_limit, new_limit_ctrl);
   }
+  // new pre_limit can push Bool/Cmp/Opaque nodes down (when one of the eliminated condition has parameters that are not
+  // loop invariant in the pre loop.
+  set_ctrl(pre_opaq, new_limit_ctrl);
+  set_ctrl(pre_end->in(1), new_limit_ctrl);
+  set_ctrl(pre_end->cmp_node(), new_limit_ctrl);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:3095]: Just for better readability, I suggest to flip those such that it matches the top down order opaque -> cmp -> bool.
```suggestion
  set_ctrl(pre_end->cmp_node(), new_limit_ctrl);
  set_ctrl(pre_end->in(1), new_limit_ctrl);
```

File: src/hotspot/share/opto/loopTransform.cpp [hunk 3]
```diff
@@ -3112,6 +3132,35 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
   // The OpaqueNode is unshared by design
   assert(opqzm->outcnt() == 1, "cannot hack shared node");
   _igvn.replace_input_of(opqzm, 1, main_limit);
+  // new main_limit can push Bool/Cmp nodes down (when one of the eliminated condition has parameters that are not loop
+  // invariant in the pre loop.
+  set_ctrl(opqzm, new_limit_ctrl);
+  set_ctrl(iffm->in(1), new_limit_ctrl);
+  set_ctrl(iffm->in(1)->in(1), new_limit_ctrl);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:3139]: Same here (flip)
```suggestion
  set_ctrl(iffm->in(1)->in(1), new_limit_ctrl);
  set_ctrl(iffm->in(1), new_limit_ctrl);
```

**Review by chhagedorn — COMMENTED** [review]
Overall, the fix looks good! A few suggestions.

**Comment by rwestrel** [comment]
@chhagedorn thanks for reviewing this. I pushed a commit that should address all your comments.

File: src/hotspot/share/opto/loopTransform.cpp [hunk 4]
```diff
@@ -2891,14 +2891,14 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
       Node *limit  = cmp->in(2);
       int scale_con= 1;        // Assume trip counter not scaled
 
-      Node *limit_c = get_ctrl(limit);
-      if (loop->is_member(get_loop(limit_c))) {
+      Node *limit_ctrl = get_ctrl(limit);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:2894]: While at it:
```suggestion
      Node* limit_ctrl = get_ctrl(limit);
```

File: src/hotspot/share/opto/loopTransform.cpp [hunk 5]
```diff
@@ -2920,25 +2920,25 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
         continue;
       }
 
-      Node *offset_c = get_ctrl(offset);
-      if (loop->is_member(get_loop(offset_c))) {
+      Node *offset_ctrl = get_ctrl(offset);
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:2923]:
```suggestion
      Node* offset_ctrl = get_ctrl(offset);
```

**Review by chhagedorn — APPROVED** [review]
Thanks for the update, looks good!

**Review by chhagedorn — APPROVED** [review]
That looks good to me, thanks!

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews

**Pull request closed without merging**